test/action_controller/serialization_scope_name_test.rb

```diff
@@ -1,63 +1,228 @@
+    end
 
-    serialization_scope :current_admin
-    before_action { request.format = :json }
+    def test_defined_serialization_scope_object
+      assert_equal @controller.view_context.class, @controller.serialization_scope.class
+    end
 
-    def current_admin
-      User.new(id: 2, name: 'Bob', admin: true)
+    def test_serialization_scope_non_admin
+      get :render_post_by_non_admin
+      expected_json = {
+        post: {
+          id: 4,
+          title: 'Title',
+          body: "The 'scope' is the 'view_context': true",
+          comments: [
+            { id: 2, body: 'Scoped' }
+          ]
+        }
+      }.to_json
+      assert_equal expected_json, @response.body
     end
 
-    def render_new_user
-      render json: User.new(id: 1, name: 'Pete', admin: false), serializer: AdminUserSerializer, adapter: :json_api
+    def test_serialization_scope_admin
+      get :render_post_by_admin
+      expected_json = {
+        post: {
+          id: 4,
+          title: 'Title',
+          body: "The 'scope' is the 'view_context': true",
+          comments: [
+            { id: 1, body: 'Admin' }
+          ]
+        }
+      }.to_json
+      assert_equal expected_json, @response.body
     end
   end
+  class NilSerializationScopeTest < ActionController::TestCase
+    class PostViewContextTestController < ActionController::Base
+      serialization_scope nil
+
+      attr_accessor :current_user
+
+      def render_post_with_no_scope
+        self.current_user = User.new(id: 3, name: 'Pete', admin: false)
+        render json: new_post, serializer: PostSerializer, adapter: :json
+      end
 
-  tests AdminUserTestController
+      # TODO: run test when
+      # global state in Serializer._serializer_instance_methods is fixed
+      # def render_post_with_passed_in_scope
+      #   self.current_user = User.new(id: 3, name: 'Pete', admin: false)
+      #   render json: new_post, serializer: PostSerializer, adapter: :json, scope: current_user, scope_name: :current_user
+      # end
+
+      private
+
+      def new_post
+        Post.new(id: 4, title: 'Title')
+      end
+    end
+    tests PostViewContextTestController
+
+    def test_nil_serialization_scope
+      assert_nil @controller._serialization_scope
+    end
+
+    def test_nil_serialization_scope_object
+      assert_nil @controller.serialization_scope
+    end
+
+    # TODO: change to NoMethodError and match 'admin?' when the
+    # global state in Serializer._serializer_instance_methods is fixed
+    def test_nil_scope
+      if Rails.version.start_with?('4.0')
+        exception_class = NoMethodError
+        exception_matcher = 'admin?'
+      else
+        exception_class = NameError
+        exception_matcher = /admin|current_user/
+      end
+      exception = assert_raises(exception_class) do
+        get :render_post_with_no_scope
+      end
+      assert_match exception_matcher, exception.message
+    end
```

Comment:
Bug in `Serializer._serializer_instance_methods` is causing the scoped method to sometimes be available to `read_attribute_for_serialization` and sometime not, since the public instance methods are cached once, but they change with scopes.